internal/graph/cursors.go

```diff
@@ -0,0 +1,267 @@
+package graph
+
+import (
+	"strconv"
+
+	"golang.org/x/exp/slices"
+
+	"github.com/authzed/spicedb/pkg/datastore/options"
+	v1 "github.com/authzed/spicedb/pkg/proto/dispatch/v1"
+	"github.com/authzed/spicedb/pkg/spiceerrors"
+	"github.com/authzed/spicedb/pkg/tuple"
+)
+
+// cursorInformation is a struct which holds information about the current incoming cursor (if any)
+// and the sections to be added to the *outgoing* partial cursor.
+type cursorInformation struct {
+	// currentCursor is the current incoming cursor. This may be nil.
+	currentCursor *v1.Cursor
+
+	// outgoingCursorSections are the sections to be added to the outgoing *partial* cursor.
+	// It is the responsibility of the *caller* to append together the incoming cursors to form
+	// the final cursor.
+	outgoingCursorSections []string
+}
+
+// newCursorInformation constructs a new cursorInformation struct from the incoming cursor (which
+// may be nil)
+func newCursorInformation(incomingCursor *v1.Cursor) cursorInformation {
+	return cursorInformation{
+		currentCursor:          incomingCursor,
+		outgoingCursorSections: nil,
+	}
+}
+
+// responsePartialCursor is the *partial* cursor to return in a response.
+func (ci cursorInformation) responsePartialCursor() *v1.Cursor {
+	return &v1.Cursor{
+		Sections: ci.outgoingCursorSections,
+	}
+}
+
+// hasPrefix returns true if the current cursor has the given name as the prefix of the cursor.
+func (ci cursorInformation) hasPrefix(name string) (bool, error) {
```

Comment:
`hasSection` might be inferred to indicate that the name exists in `sections` anywhere; I wanted to be explicit

Reply:
Ok, I see your point. You seem to have used for that the term `head` elsewhere, WDYT of `hasHeadSection`.
I say this because as I was trying to parse all of this code, the concept of "prefix" didn't show up elsewhere and it tripped me.